README.md
Porting of the Kodi plugin to xbmc4xbox



•To do list:

Live radio (out of memory, works if you first download and run the m3u8 file )

Live TV (out of memory, need to understand how it works)

Fix some on demand programs not playing (the m3u8 file works, but on the Xbox it searches for a different link, need to look at it)



CHANGELOG.txt
[B]0.3.0[/B]
Removed script.common.plugin.cache, now the plugin runs a lot faster (from 20/25 seconds to 3 seconds between menues)
[B]0.2.0[/B]
Fixes to all the thumbs, now all other them can be (hopefully) retrieved, with this all the menus are now working (it's the second time I say that, I know)
[B]0.1.0[/B]
New versioning, now can retrive thumbs from the net, disabled artworks (use too much memory), radio replay now works, same for short programs (on the On Demand section), all the menues should work now. (with upgraded ram/cpu i think you can watch every on demand program, need testing)


resources/lib/tgr.py
# -*- coding: utf-8 -*-
import sys
from xml.dom import minidom
import resources.lib.utils as utils

import urllib2


class TGR:
    _baseurl = "http://www.tgr.rai.it"
    
    def getProgrammes(self):
        url = "http://www.tgr.rai.it/dl/tgr/mhp/home.xml"
        try:
            xmldata = urllib2.urlopen(url).read()
            xmldata = utils.checkStr(xmldata)
            dom = minidom.parseString(xmldata)

            programmes = []
            for node in dom.getElementsByTagName('item'):
                item = {}
                # behaviour = {region, select, list}
                item["behaviour"] = node.attributes["behaviour"].value
                item["title"] = node.getElementsByTagName('label')[0].childNodes[0].data
                for url in  node.getElementsByTagName('url'):
                    if url.attributes["type"].value == "image":
                        item["image"] = self._baseurl + url.childNodes[0].data
                    elif url.attributes["type"].value == "list":
                        item["url"] = self._baseurl + url.childNodes[0].data
                programmes.append(item)

            return programmes
        
        except urllib2.HTTPError:
            return []
     
    def getList(self, url):
        
        try:
            xmldata = urllib2.urlopen(url).read()
            xmldata = utils.checkStr(xmldata)
            dom = minidom.parseString(xmldata)

            items = []
            for node in dom.getElementsByTagName('item'):
                item = {}
                # behaviour = {list, video}
                item["behaviour"] = node.attributes["behaviour"].value
                item["title"] = node.getElementsByTagName('label')[0].childNodes[0].data
                for url in  node.getElementsByTagName('url'):
                    foundUrl = False
                    if url.attributes["type"].value == "list":
                        item["url"] = self._baseurl + url.childNodes[0].data
                        foundUrl = True
                    elif url.attributes["type"].value == "video":
                        if url.attributes["type"].value == "video":
                            item["url"] = url.childNodes[0].data
                            foundUrl = True
                    if foundUrl:
                        items.append(item)

            return items

        except urllib2.HTTPError:
            return []
        


resources/lib/search.py
# -*- coding: utf-8 -*-
import sys
import json
import urllib
import resources.lib.utils as utils

import urllib2


class Search:
    baseUrl = "http://www.rai.it"

    newsArchives = {"TG1": "NomeProgramma:TG1^Tematica:Edizioni integrali",
        "TG2": "NomeProgramma:TG2^Tematica:Edizione integrale",
        "TG3": "NomeProgramma:TG3^Tematica:Edizioni del TG3"}
    
    newsProviders = {"TG1": "Tematica:TG1",
        "TG2": "Tematica:TG2",
        "TG3": "Tematica:TG3",
        "Rai News": "Tematica:Rai News",
        "Rai Sport": "Tematica:spt",
        "Rai Parlamento": "PageOB:Page-f3f817b3-1d55-4e99-8c36-464cea859189"}

    tematiche = ["Attualità", "Bianco e Nero", "Cinema", "Comici", "Cronaca", "Cucina", "Cultura", "Cultura e Spettacoli", "Economia", "Fiction",
        "Hi tech", "Inchieste", "Incontra", "Interviste", "Istituzioni", "Junior", "Moda", "Musica", "News", "Politica", "Promo", "Reality",
        "Salute", "Satira", "Scienza", "Società", "Spettacolo", "Sport", "Storia", "Telefilm", "Tempo libero", "Viaggi"]
    
    def getLastContentByTag(self, tags="", numContents=16):
        try: tags = urllib.quote(tags)
        except: tags = urllib.parse.quote(tags)
        domain = "RaiTv"
        xsl = "rai_tv-statistiche-raiplay-json"
        
        url = self.baseUrl +  "/StatisticheProxy/proxyPost.jsp?action=getLastContentByTag&numContents=%s&tags=%s&domain=%s&xsl=%s" % \
              (str(numContents), tags, domain, xsl)
        response = json.loads(utils.checkStr(urllib2.urlopen(url).read()))
        return response["list"]
    
    def getMostVisited(self, tags, days=7, numContents=16):
        try: tags = urllib.quote(tags)
        except: tags = urllib.parse.quote(tags)
        domain = "RaiTv"
        xsl = "rai_tv-statistiche-raiplay-json"
        
        url = self.baseUrl +  "/StatisticheProxy/proxyPost.jsp?action=mostVisited&days=%s&state=1&records=%s&tags=%s&domain=%s&xsl=%s" % \
            (str(days), str(numContents), tags, domain, xsl)
        response = json.loads(utils.checkStr(urllib2.urlopen(url).read()))
        return response["list"]


resources/lib/relinker.py
# -*- coding: utf-8 -*-
import sys
import urllib
import resources.lib.utils as utils

import urllib2
import urlparse
from urllib import urlencode

    
class Relinker:
    # Firefox 52 on Android
    # UserAgent = "Mozilla/5.0 (Android; Mobile; rv:52.0) Gecko/52.0 Firefox/52.0"
    # Firefox 52 ESR on Windows 7
    # UserAgent = "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:52.0) Gecko/20100101 Firefox/52.0"
    # Firefox 52 ESR on Linux
    # UserAgent = "Mozilla/5.0 (X11; Linux x86_64; rv:52.0) Gecko/20100101 Firefox/52.0"
    # Chrome 64 on Windows 10
    # UserAgent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/64.0.3282 Safari/537.36"
    # Raiplay android app
    UserAgent = "Android 4.2.2 (smart) / RaiPlay 2.1.3 / WiFi"
    
    def __init__(self):
        opener = urllib2.build_opener()
        # Set User-Agent
        opener.addheaders = [('User-Agent', self.UserAgent)]
        urllib2.install_opener(opener)

    def getURL(self, url):
        scheme, netloc, path, params, query, fragment = urlparse.urlparse(url)
        qs = urlparse.parse_qs(query)
    
        # output=20 url in body
        # output=23 HTTP 302 redirect
        # output=25 url and other parameters in body, space separated
        # output=44 XML (not well formatted) in body
        # output=47 json in body
        # pl=native,flash,silverlight
        # A stream will be returned depending on the UA (and pl parameter?)
        
        if "output" in qs:
            del(qs['output'])
        qs['output'] = "20"
        
        query = urlencode(qs, True)
        url = urlparse.urlunparse((scheme, netloc, path, params, query, fragment))
                
        try:
            response = utils.checkStr(urllib2.urlopen(url).read())
            mediaUrl = response.strip()

            # Workaround to normalize URL if the relinker doesn't
            try: 
                mediaUrl = urllib.quote(mediaUrl, safe="%/:=&?~#+!$,;'@()*[]")
            except: 
                mediaUrl = urllib.parse.quote(mediaUrl, safe="%/:=&?~#+!$,;'@()*[]")
            return mediaUrl

        except HTTPError:
            return ''


resources/lib/raiplayradio.py
# -*- coding: utf-8 -*-
import sys
import json
import unicodedata
import resources.lib.utils as utils

import urllib2

class RaiPlayRadio:
    # Raiplay android app
    UserAgent = "okhttp/3.9.1"
    MediapolisUserAgent = "Radio/3.0.2 (Linux;Android 4.2.2) ExoPlayerLib/2.5.4"
    
    noThumbUrl = "http://www.raiplayradio.it/dl/components/img/radio/player/placeholder_img.png"
    
    # http://www.rai.it/dl/rairadio/mobile/config/RaiRadioConfig.json
    baseUrl = "http://www.raiplayradio.it/"
    channelsUrl = "http://www.raiplayradio.it/dirette/ContentSet-8e2a3414-bbfd-4a72-852b-7d827dc00e7e.html?json"
    localizeUrl = "http://mediapolisgs.rai.it/relinker/relinkerServlet.htm?cont=201342"
    palinsestoUrl = "http://www.raiplayradio.it/dl/palinsesti/Page-a47ba852-d24f-44c2-8abb-0c9f90187a3e-json.html?canale=[nomeCanale]&giorno=[dd-mm-yyyy]&mode=light"
    AzTvShowPath = "/dl/RaiTV/RaiRadioMobile/Prod/Config/programmiAZ-elenco.json"
    
    def __init__(self):
        opener = urllib2.build_opener()
        # Set User-Agent
        opener.addheaders = [('User-Agent', self.UserAgent)]
        urllib2.install_opener(opener)
    
    def getCountry(self):
        try:
            response = utils.checkStr(urllib2.urlopen(self.localizeUrl).read())
        except urllib2.HTTPError:
            response = "ERROR"
        return response
        
    def getChannels(self):
        response = json.loads(utils.checkStr(urllib2.urlopen(self.channelsUrl).read()))
        return response["dirette"]
        
    def getProgrammes(self, channelName, epgDate):
        channelTag = channelName.replace(" ", "")
        channelTag = unicodedata.normalize('NFD', channelTag).encode('ascii', 'ignore')
        url = self.palinsestoUrl
        url = url.replace("[nomeCanale]", channelTag)
        url = url.replace("[dd-mm-yyyy]", epgDate)
        response = json.loads(utils.checkStr(urllib2.urlopen(url).read()))
        return response[channelName][0]["palinsesto"][0]["programmi"]
    
    def getAudioMetadata(self, pathId):
        url = self.getUrl(pathId)
        response = json.loads(utils.checkStr(urllib2.urlopen(url).read()))
        return response["audio"]
    
    def getUrl(self, pathId):
        pathId = pathId.replace(" ", "%20")
        if pathId[0:2] == "//":
            url = "http:" + pathId
        elif pathId[0] == "/":
            url = self.baseUrl[:-1] + pathId
        else:
            url = pathId
        return url
        
    def getThumbnailUrl(self, pathId):
        if pathId == "":
            url = self.noThumbUrl
        else:
            url = self.getUrl(pathId)
            url = url.replace("[RESOLUTION]", "256x-")
        return url
 

resources/lib/raiplay.py
# -*- coding: utf-8 -*-
import sys
import json
import re
import xbmc
import resources.lib.utils as utils

import urllib2
import HTMLParser


class RaiPlay:
    # Raiplay android app
    UserAgent = "Dalvik/1.6.0 (Linux; U; Android 4.2.2; GT-I9105P Build/JDQ39)"
    MediapolisUserAgent = "Android 4.2.2 (smart) / RaiPlay 2.1.3 / WiFi"
    
    noThumbUrl = "http://www.rai.it/dl/components/img/imgPlaceholder.png"
    
    # From http://www.raiplay.it/mobile/prod/config/RaiPlay_Config.json
    baseUrl = "https://www.raiplay.it/"
    channelsUrl = "http://www.rai.it/dl/RaiPlay/2016/PublishingBlock-9a2ff311-fcf0-4539-8f8f-c4fee2a71d58.html?json"
    localizeUrl = "http://mediapolisgs.rai.it/relinker/relinkerServlet.htm?cont=201342"
    menuUrl = "http://www.rai.it/dl/RaiPlay/2016/menu/PublishingBlock-20b274b1-23ae-414f-b3bf-4bdc13b86af2.html?homejson"
    palinsestoUrl = "https://www.raiplay.it/palinsesto/app/old/[nomeCanale]/[dd-mm-yyyy].json"
    palinsestoUrlHtml = "https://www.raiplay.it/palinsesto/guidatv/lista/[idCanale]/[dd-mm-yyyy].html"
    onAirUrl = "https://www.raiplay.it/palinsesto/onAir.json" 
    AzTvShowPath = "/dl/RaiTV/RaiPlayMobile/Prod/Config/programmiAZ-elenco.json"
    
    # Rai Sport urls
    RaiSportMainUrl = 'https://www.raisport.rai.it'
    RaiSportLiveUrl= RaiSportMainUrl + '/dirette.html'
    RaiSportArchivioUrl = RaiSportMainUrl + '/archivio.html'        
    RaiSportSearchUrl = RaiSportMainUrl +  "/atomatic/news-search-service/api/v1/search?transform=false"
    RaiPlayAddonHandle = None
    
    def __init__(self, addonHandle):
        opener = urllib2.build_opener()
        # Set User-Agent
        opener.addheaders = [('User-Agent', self.UserAgent)]
        urllib2.install_opener(opener)
        self.RaiPlayAddonHandle = addonHandle
        
    def getCountry(self):
        try:
            response = utils.checkStr(urllib2.urlopen(self.localizeUrl).read())
        except urllib2.HTTPError:
            response = "ERROR"
        return response
        
    def getChannels(self):
        response = json.loads(utils.checkStr(urllib2.urlopen(self.channelsUrl).read()))
        return response["dirette"]
    
    def getOnAir(self):
        response = json.loads(utils.checkStr(urllib2.urlopen(self.onAirUrl).read()))
        return response["on_air"]
    
    def getHomePage(self):
        response = json.loads(utils.checkStr(urllib2.urlopen(self.baseUrl + 'index.json').read()))
        return response["contents"]
      
    def getRaiSportLivePage(self):
        chList = []
        try:
            response = utils.checkStr(urllib2.urlopen(self.RaiSportLiveUrl).read())
        except urllib2.HTTPError:
            response = ''
            
        m = re.search('<ul class="canali">(?P<list>.*)</ul>', response, re.S)
        if m:
            channels = re.findall ('<li>(.*?)</li>', m.group('list'), re.S)
            for ch in channels:
                url = re.search('''data-video-url=['"](?P<url>[^'^"]+)['"]''', ch)
                if url:
                    url = url.group('url')
                    icon = re.search('''stillframe=['"](?P<url>[^'^"]+)['"]''', ch )
                    if icon:
                        icon = self.getUrl(icon.group('url'))
                    else:
                        icon = ''
                    title = re.search(">(?P<title>[^<]+)</a>", ch )
                    if title:
                        title = title.group('title')
                    else:
                        title = self.RaiPlayAddonHandle.getLocalizedString(32014)
                    chList.append({'title':title, 'url':url, 'icon':icon})
        
        return chList
    
    def fillRaiSportKeys(self):
        # search for items in main menu
        RaiSportKeys=[]
        
        try:        
            data = utils.checkStr(urllib2.urlopen(self.RaiSportMainUrl).read())
        except urllib2.HTTPError:
            data = ''
        
        m = re.search("<a href=\"javascript:void\(0\)\">Menu</a>(.*?)</div>", data,re.S)
        if not m: 
            return []
        menu = m.group(0)
        
        links = re.findall("<a href=\"(?P<url>[^\"]+)\">(?P<title>[^<]+)</a>", menu)
        good_links=[]
        for l in links:
            if ('/archivio.html?' in l[0]) and not ('&amp;' in l[0]):
                good_links.append({'title': l[1], 'url' : l[0]})
        
        good_links.append({'title': self.RaiPlayAddonHandle.getLocalizedString(32015), 'url' : '/archivio.html?tematica=altri-sport'})
        
        # open any single page in list and grab search keys
        
        for l in good_links:
            try:
                data = utils.checkStr(urllib2.urlopen(self.RaiSportMainUrl + l['url']).read())
            except urllib2.HTTPError:
                data = ''

            dataDominio= re.findall("data-dominio=\"(.*?)\"", data)
            dataTematica = re.findall("data-tematica=\"(.*?)\"", data)
            xbmc.log(str(dataTematica))
            if dataTematica:
                if len(dataTematica) > 1:
                    del(dataTematica[0])
        
                try:
                    title=dataTematica[0].split('|')[0]
                    title = utils.checkStr(HTMLParser.HTMLParser().unescape(title))
                    params={'title': title, 'dominio': dataDominio[0], 'sub_keys' : dataTematica}
                
                    RaiSportKeys.append(params)
                except:
                    xbmc.log("error in key %s" % str(dataTematica))
        
        return RaiSportKeys
    
    def getRaiSportVideos(self, key, domain, page):
        videos = []
        header = {
                  'Accept': 'application/json, text/javascript, */*; q=0.01' ,
                  'Content-Type': 'application/json; charset=UTF-8',
                  'Origin': 'https://www.raisport.rai.it',
                  'Referer': 'https://www.raisport.rai.it/archivio.html',
                  'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.132 Safari/537.36',
                  'X-Requested-With': 'XMLHttpRequest',
                 }
        page = int(page)
        pageSize = 50

        payload = {
            "page": page,
            "pageSize": pageSize,
            "filters":{
                "tematica":[key],
                "dominio": domain
            }
        }
        postData=json.dumps(payload)
        try:
            req = urllib2.Request(self.RaiSportSearchUrl, postData, header)
            response = urllib2.urlopen(req)
        except TypeError:
            req = urllib2.Request(self.RaiSportSearchUrl, postData.encode('utf-8'), header)
            response = urllib2.urlopen(req)
        
        if response.code != 200:
            return []
        
        data = utils.checkStr(response.read())
        j = json.loads(data)
        
        if 'hits' in j:
            h = j['hits']
            if 'hits' in h:
                for hh in h['hits']:
                    if '_source' in hh:
                        news_type = hh['_source']['tipo']
                        if news_type == 'Video' and 'media' in hh['_source']:
                            relinker_url = hh['_source']['media']['mediapolis']

                            if 'durata' in hh['_source']['media']:
                                d= hh['_source']['media']['durata'].split(":")
                                duration = int(d[0])*3600 + int(d[1])*60 + int(d[2])
                            else:
                                duration = 0

                            icon = self.RaiSportMainUrl + hh['_source']['immagini']['default']
                            title = hh['_source']['titolo']
                            creation_date = hh['_source']['data_creazione']
                            if 'sommario' in hh['_source']: 
                                desc = creation_date + '\n' + hh['_source']['sommario']
                            else:
                                desc = creation_date 

                            params= {'mode':'raisport_video', 'title': title, 'url': relinker_url, 'icon': icon, 
                                     'duration' : duration, 'aired': creation_date, 'desc': desc}
                            videos.append(params)

            if h['total'] > (page + pageSize):
                page += pageSize
                params = {'mode':'raisport_subitem', 'title': xbmc.getLocalizedString(33078), 'page': page}
                videos.append(params)
        
        return videos
    
    def getProgrammes(self, channelName, epgDate):
        channelTag = channelName.replace(" ", "")
        url = self.palinsestoUrl
        url = url.replace("[nomeCanale]", channelTag)
        url = url.replace("[dd-mm-yyyy]", epgDate)
        response = json.loads(utils.checkStr(urllib2.urlopen(url).read()))
        try:
            key = channelName if channelName in response else ''
            oRetVal = response[key][0]["palinsesto"][0]["programmi"]
        except:
            oRetVal = None
        return oRetVal
    
    def getProgrammesHtml(self, channelName, epgDate):
        channelTag = channelName.replace(" ", "-").lower()
        url = self.palinsestoUrlHtml
        url = url.replace("[idCanale]", channelTag)
        url = url.replace("[dd-mm-yyyy]", epgDate)
        try:
            data = utils.checkStr(urllib2.urlopen(url).read())
        except urllib2.HTTPError:
            data = ''
        return data
        
    def getMainMenu(self):
        response = json.loads(utils.checkStr(urllib2.urlopen(self.menuUrl).read()))
        return response["menu"]

    # RaiPlay Genere Page
    # RaiPlay Tipologia Page
    def getCategory(self, pathId):
        url = self.getUrl(pathId)
        try:
            response = json.loads(utils.checkStr(urllib2.urlopen(url).read()))
            return response["contents"]
        except: 
            return []
  
    # Raiplay Tipologia Item
    def getProgrammeList(self, pathId):
        url = self.getUrl(pathId)
        try:
            response = json.loads(utils.checkStr(urllib2.urlopen(url).read()))
            return response["contents"]
        except:
            return []

    # Raiplay AZ List
    def getProgrammeListOld(self, pathId):
        url = self.getUrl(pathId)
        try:
            response = json.loads(utils.checkStr(urllib2.urlopen(url).read()))
            return response
        except:
            return []
      
    #  PLR programma Page
    def getProgramme(self, pathId):
        url = self.getUrl(pathId)
        try:
            response = json.loads(utils.checkStr(urllib2.urlopen(url).read()))
            return response
        except:
            return[]
        
    def getContentSet(self, url):
        url = self.getUrl(url)
        try:
            response = json.loads(utils.checkStr(urllib2.urlopen(url).read()))
            return response["items"]
        except:
            return []
    
    def getVideoMetadata(self, pathId):
        url = self.getUrl(pathId)
        if url.endswith(".html"):
            url = url.replace(".html",".json")
            
        try:
            response = json.loads(utils.checkStr(urllib2.urlopen(url).read()))
            return response["video"]
        except:
            return []
    
    def getIndexFromJSON(self, pathId):
        url = self.getUrl(pathId)
        try:
            response = json.loads(utils.checkStr(urllib2.urlopen(url).read()))
        
            index = []
            for i in response["contents"]:
              if len(response["contents"][i])>0:
                index.append(i)
        
            index.sort()
            return index
        except:
            return []
    
    def getUrl(self, pathId):
        url = pathId.replace(" ", "%20")
        
        if url.startswith("/raiplay/"):
            url = url.replace("/raiplay/","https://raiplay.it/")
        
        if url[0:2] == "//":
            url = "https:" + url
        elif url[0] == "/":
            url = self.baseUrl[:-1] + url
        
        # fix old format of url for json
        if url.endswith(".html?json"):
            url = url.replace(".html?json", ".json")
        elif url.endswith("/?json"):
            url = url.replace("/?json","/index.json")
        elif url.endswith("?json"):
            url = url.replace("?json",".json")
        
        return url
        
    def getThumbnailUrl(self, pathId):
        if pathId == "":
            url = self.noThumbUrl
        else:
            url = self.getUrl(pathId)
            url = url.replace("[RESOLUTION]", "256x-")
        return url
 
